src/site/content/en/blog/more-capable-form-controls/index.md

```diff
@@ -0,0 +1,305 @@
+
+  // The following properties and methods aren't strictly required, 
+  // but native form controls provide them. Providing them helps
+  // ensure consistency with native controls.
+  get form() { return this._internals.form; }
+  get name() { return this.getAttribute('name'); }
+  get type() { return this.localName; }
+  get validity() {return this.internals_.validity; }
+  get validationMessage() {return this.internals_.validationMessage; }
+  get willValidate() {return this.internals_.willValidate; }
+
+  checkValidity() { return this.internals_.checkValidity(); }
+  reportValidity() {return this.internals_.reportValidity(); }
+
+  ... 
+}
+customElements.register('my-counter', MyCounter);
+```
+
+Once registered, you can use this element wherever you'd use a native form control:
+
+```html
+<form> 
+  <label>Number of bunnies: <my-counter></my-counter></label>
+  <button type="submit">Submit</button>
+</form>
+```
+
+### Setting a value {: #setting-a-value }
+
+The `attachInternals` method returns an `ElementInternals` object that provides access to form control APIs. The most basic of these is the `setFormValue` method, which sets the current value of the control. 
+
+The `setFormValue` method can take one of three types of values:
+
+*   A string value.
+*   A [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) object.
+*   A [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) object. You can use a `FormData` object to pass multiple values (for example, a credit card input control might pass a card number, expiration date, and verification code).
+
+To set a simple value:
+
+```js
+this._internals.setFormValue(this._value);
+```
+
+To set multiple values, you can do something like this:
+
+```js
+// Use the control's name as the base name for submitted data
+const n = this.getAttribute('name');
+const entries = new FormData();
+entries.append(n + '-first-name', this.firstName_);
+entries.append(n + '-last-name', this.lastName_);
+this.internals_.setFormValue(entries);
+```
+
+{% Aside %}The `setFormValue` method takes a second, optional `state` parameter, used to store the internal state of the control. For more information, see [Restoring form state](#restoring-form-state).{% endAside %}
+
+### Input validation
+
+Your control can also participate in form validation by calling the `setValidity`
+method on the internals object.
+
+```js
+// Assume this is called whenever the internal value is updated
+onUpdateValue() {
+  // 
+  if (!this.matches(':disabled') && this.hasAttribute('required') &&
+      this._value < 0) {
+    this._internals.setValidity({customError: true}, 'Value cannot be negative.');
+  } 
+  else {
+    this._internals.setValidity({});
+  }
+  this.internals.setFormValue(this._value);
+  }
+```
+
+You can style a form-associated custom element with the `:valid` and `:invalid`
+pseudoclasses, just like a built-in form control.
+
+### Lifecycle callbacks
```

Comment:
I was curious what you think about formatting the section like this:

```
### Lifecycle callbacks

A form-associated custom element API includes a set of extra lifecycle callbacks to tie in to the form lifecycle. The callbacks are optional: only implement a callback if your element needs to do something at that point in the lifecycle.

#### `formAssociatedCallback(form)`

Called when the browser associates the element with a form element, or disassociates the element from a form element.

#### `formDisabledCallback(disabled)`

Called after the `disabled` state of the element changes, either because the `disabled` attribute of this element was added or removed; or because the `disabled` state changed on a `<fieldset>` that's an ancestor of this element. The `disabled` parameter represents the new [disabled state](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-disabled) of the element. The element may, for example, disable elements in its shadow DOM when it is disabled.

#### `formResetCallback()`

Called after the form is reset. The element should reset itself to some kind of default state. For native inputs, this usually involves setting the `value` property to match the `value` attribute set in markup (or in the case of a checkbox, setting the `checked` property to match the `checked` attribute.

#### `formStateRestoreCallback(state, mode)`

Called in one of two circumstances:

* When the browser restores the state of the element (for example, after a navigation, or when the browser restarts).  The <code>mode</code> argument is <code>"restore"</code> in this case.
* When the browser's input-assist features such as form autofilling sets a value. The <code>mode</code> argument is <code>"autocomplete"</code> in this case.

The type of the first argument depends on how the <code>setFormValue</code> method was called. For more details, see [Restoring form state](#restoring-form-state).
```

It'll look something like this:

![image](https://user-images.githubusercontent.com/1066253/62668841-3fbf6a00-b942-11e9-8392-4a4fce50b872.png)

Reply:
Hm... Yeah, I could do that. I had a variation of that, but the headers kind of duplicated the callback names. This could work. 

It's a little weird because there's another H4 after the callbacks called "Restoring form state." It's the gnarly details of how formStateRestoreCallback interacts with setFormValue. It felt like too much information to cram into the callback section. I'll take another look at it.

Reply:
Done.